File: packages/adapter-langchain/src/langgraph.ts
import { BaseCallbackHandler } from "@langchain/core/callbacks/base";
import type { FlightRecorder } from "@afr/sdk";
import { getDeterministicStateHash, computeStateDiff } from "@afr/contracts";

export interface LangGraphHandlerOptions {
  recorder: FlightRecorder;
}

export function createLangGraphHandler(options: LangGraphHandlerOptions) {
  return new LangGraphCallbackHandler(options.recorder);
}

export class LangGraphCallbackHandler extends BaseCallbackHandler {
  name = "LangGraphCallbackHandler";
  
  private stateMap = new Map<string, Record<string, any>>();

  constructor(private recorder: FlightRecorder) {
    super();
  }

  async handleChainStart(
    chain: any,
    inputs: any,
    runId: string,
    parentRunId?: string
  ) {
    // If it is the main graph execution, we can capture the initial state
    if (chain?.id && chain.id[chain.id.length - 1] === "Graph") {
      const initialState = typeof inputs === "object" ? JSON.parse(JSON.stringify(inputs)) : {};
      this.stateMap.set(runId, initialState);
      
      const stateHash = getDeterministicStateHash(initialState);
      await this.recorder.stateSnapshot({
        parentSpanId: parentRunId || null,
        spanId: runId,
        nodeName: "Graph__start",
        stateHash,
        stateDiff: initialState,
        fullState: initialState
      });
    }
  }

  async handleChainEnd(outputs: any, runId: string) {
    // In LangGraph, the outputs of a node are typically the state updates
    if (outputs && typeof outputs === "object" && !Array.isArray(outputs)) {
      const currentState = this.stateMap.get(runId) || {};
      const { stateDiff, removedKeys } = computeStateDiff(currentState, outputs);
      
      if (Object.keys(stateDiff).length > 0 || removedKeys.length > 0) {
        // Deep clone before emission to protect against mutations
        const nextState = Object.freeze(JSON.parse(JSON.stringify({ ...currentState, ...outputs })));
        this.stateMap.set(runId, nextState);
        
        const stateHash = getDeterministicStateHash(nextState);
        
        await this.recorder.stateSnapshot({
          parentSpanId: runId,
          nodeName: "node_update",
          stateHash,
          stateDiff,
          removedKeys,
          fullState: nextState
        });
      }
    }
  }

  async handleToolEnd(output: any, runId: string) {
    // No-op for now. 
    // We let normal tools emit their tool result via the main handler.
  }

  async handleCustomEvent(
    eventName: string,
    data: any,
    runId: string,
    _tags?: string[],
    _metadata?: Record<string, any>
  ) {
    if (eventName === "langgraph.checkpoint" || eventName === "langgraph.interrupt" || eventName === "on_custom_event") {
      const isInterrupt = eventName === "langgraph.interrupt" || data?.__interrupt;
      if (isInterrupt || data?.checkpoint) {
        const currentState = this.stateMap.get(runId) || {};
        await this.recorder.checkpoint({
          parentSpanId: runId,
          checkpointId: data?.id || runId,
          reason: data?.reason || data?.name || "interrupt",
          state: currentState
        });
      }
    }
  }
}

File: python/aerograph-langchain/src/aerograph_langchain/langgraph.py
import uuid
from typing import Any, Dict, List, Optional

from aerograph_sdk.events import build_state_snapshot_event, build_checkpoint_event
from aerograph_sdk.contracts.generated import StateSnapshotEvent, CheckpointEvent
from aerograph_langchain.span_ids import derive_span_id


def map_state_snapshot(
    run_id: uuid.UUID,
    trace_id: str,
    node_name: str,
    state_hash: str,
    state_diff: Dict[str, Any],
    full_state: Dict[str, Any],
    removed_keys: Optional[List[str]] = None,
    parent_run_id: Optional[uuid.UUID] = None,
) -> StateSnapshotEvent:
    span_id = derive_span_id(run_id)
    parent_span_id = derive_span_id(parent_run_id) if parent_run_id else None

    return build_state_snapshot_event(
        trace_id=trace_id,
        span_id=span_id,
        parent_span_id=parent_span_id,
        actor_id="langgraph",
        actor_name="LangGraph",
        node_name=node_name,
        state_hash=state_hash,
        state_diff=state_diff,
        full_state=full_state,
        removed_keys=removed_keys or [],
    )


def map_checkpoint(
    run_id: uuid.UUID,
    trace_id: str,
    checkpoint_id: str,
    reason: str,
    state: Dict[str, Any],
    parent_run_id: Optional[uuid.UUID] = None,
) -> CheckpointEvent:
    span_id = derive_span_id(run_id)
    parent_span_id = derive_span_id(parent_run_id) if parent_run_id else None

    return build_checkpoint_event(
        trace_id=trace_id,
        span_id=span_id,
        parent_span_id=parent_span_id,
        actor_id="langgraph",
        actor_name="LangGraph",
        checkpoint_id=checkpoint_id,
        reason=reason,
        state=state,
    )
